File: client/src/types/HousingUnit.ts
import { Feature } from "./GeoDataResponse";
import { Landlord } from "./Landlord";
export class HousingUnit {
  lat: number;
  lng: number;
  lst_rented?: string;
  unit?: string;
  unitId: string;
  streetNumber: string;
  streetName: string;
  postalCode: string;
  city: string = "NYC";
  county: string;
  country: string = "USA";
  cr_ts: string;
  landlord: Landlord;
  label: string;

  toBackend = (feature: Feature) => {
    this.lat = feature.geometry.coordinates[1];
    this.lng = feature.geometry.coordinates[0];
    this.streetNumber = feature.properties.housenumber;
    this.streetName = feature.properties.street;
    this.postalCode = feature.properties.postalcode;
    this.county = feature.properties.borough;
    return this;
  };

  fromBackend = (housingUnit: HousingUnit) => {
    this.city = housingUnit.city;
    this.country = housingUnit.country;
    this.county = housingUnit.county;
    this.cr_ts = housingUnit.cr_ts;
    this.landlord = housingUnit.landlord;
    this.lat = housingUnit.lat;
    this.lng = housingUnit.lng;
    this.lst_rented = housingUnit.lst_rented;
    this.postalCode = housingUnit.postalCode;
    this.streetNumber = housingUnit.streetNumber;
    this.streetName = housingUnit.streetName;
    this.unit = housingUnit.unit;
    this.unitId = housingUnit.unitId;
    this.lat = housingUnit.lat;
    this.lng = housingUnit.lng;
    this.label = this.createLabel();
    return this;
  };

  createLabel = () => {
    return `<ul>
			<li>address: ${this.streetNumber} ${this.streetName}, ${this.county}</li>
			<li>landlord name: ${this.landlord}</li>
			</ul>`;
  };

  //   getLandlordInfo = () => {
  // 	if(!this.landlord) return;
  // 	const htmlString = ""
  // 	if(this.landlord.name)
  //   }
}

File: server/src/main/java/com/vouchapp/vouch/model/HousingUnit.java
package com.vouchapp.vouch.model;


import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.GenericGenerator;
import org.springframework.validation.annotation.Validated;

import java.sql.Timestamp;
import java.util.Date;
import java.util.List;

//uuid (can be zipcode+random digits)	Create Time Stamp	Last rented	Cost?	Street Number	Unit number	Street Name	Zipcode	City	County	Country	Neighborhood	LandlordID
@Entity(name = "units")
@Data
@Validated
@NoArgsConstructor
@Table(uniqueConstraints = { @UniqueConstraint(columnNames = { "lat", "lng" }) })
public class HousingUnit {

    private static final List<String> requireFields = List.of("lat", "lng", "streetNumber", "streetName", "postalCode", "city", "county", "country");

    @Id
    @GeneratedValue(generator = "unitId-generator")
    @GenericGenerator(name = "unitId-generator",
            strategy = "com.vouchapp.vouch.model.generator.UnitIdGenerator")
    private String unitId;

    @Column(nullable = false, updatable = false)
    @CreationTimestamp
    private Timestamp cr_ts;

    @NotNull
    @Column(nullable = false)
    private Number lat;
    @NotNull
    @Column(nullable = false)
    private Number lng;

    private Date lst_rented;
    private String unit;

    @NotNull
    @Column(nullable = false)
    private String streetNumber;

    @NotNull
    @Column(nullable = false)
    private String streetName;

    @NotNull
    @Column(nullable = false)
    private String postalCode;
    @NotNull
    @Column(nullable = false)
    private String city;
    @NotNull
    @Column(nullable = false)
    private String county;
    @NotNull
    @Column(nullable = false)
    private String country;

    @ManyToOne
    @JoinColumn(name = "landlordId")
    private Landlord landlord;

    public HousingUnit(String unit, @NotNull Number lat, @NotNull Number lng, @NotNull String streetNumber, @NotNull String streetName, @NotNull String postalCode, @NotNull String city, @NotNull String county, @NotNull String country) {
//        this.lst_rented = lastRented;
        this.unit = unit;
        this.lat = lat;
        this.lng = lng;
        this.streetNumber = streetNumber;
        this.streetName = streetName;
        this.postalCode = postalCode;
        this.city = city;
        this.county = county;
        this.country = country;
    }
}


